### PR TITLE
Enabling exponentiation operator 

### PR DESCRIPTION
**Pull request by abchatra**
Update as per spec to  exponentiation operator with new grammar as in 12.6.  Enabling it by default as well.

### CODE REVIEW

File: test/es6/expo.js [hunk 1]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
+            
         }
     },
     {
         name: "Exponentiation operator must be evaluated before multiplication ",
         body: function () {
-            assert.areEqual((2 * 3) ** 2, 2 * 3 ** 2, "Exponentiation operator has lower precedence than multiplicative operator *");
+            assert.areEqual(2 * (3 ** 4), 2 * 3 ** 4, "Exponentiation operator has higher precedence than multiplicative operator *");
+            assert.areEqual(2 * (3 ** 4) * 5 , 2 * 3 ** 4 * 5, "Exponentiation operator has higher precedence than multiplicative operator *");            
             assert.areEqual(2 + (3 ** 2), 2 + 3 ** 2, "Exponentiation operator has higher precedence than additive operator +");
```
Comment (goyakin) [inline test/es6/expo.js]: Rather than using the operator itself for the expected values, I think we should use `Math.pow` instead.

Comment (ianwjhalliday) [inline test/es6/expo.js]: Well in this case we're testing order of operations, so the expecteds show the expected precedence.  I think this is okay given that there are ** tests earlier.

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -7821,9 +7821,17 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
...
             {
-                DeferOrEmitPotentialSpreadError(pnodeT);
+                if (!fAllowEllipsis)
+                {
+                    DeferOrEmitPotentialSpreadError(pnodeT);
+                }
+            }
+            else if (m_token.tk == tkExpo)
+            {
+                //Unary operator on the left handside of ** is unexpected, except ++, -- or ...
```
Comment (akroshg) [inline lib/Parser/Parse.cpp]: how '...' operator be used with ** ?

**Comment by akroshg** [comment]
Looks good to me

File: lib/Parser/Parse.cpp (hunk 2 above)
Comment (ljharb) [inline lib/Parser/Parse.cpp]: :trollface: 
```
Number.prototype[Symbol.iterator] = function* () {
  for (let i = 0, i < this; ++i) { yield i; }
};
[...3**4]
```

File: test/es6/expo.js [hunk 3]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
         }
     },
+    {
+        name: "Exponentiation syntax error cases ",
+        body: function () {
+            assert.throws(function () { eval("-2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator +", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("+2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator -", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("delete 2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator delete", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("!2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator !", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
```
Comment (ianwjhalliday) [inline test/es6/expo.js]: nit: in the description it should be `a union` not `an union`

File: test/es6/expo.js [hunk 4]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
+            assert.areEqual(2 * (3 ** 4), 2 * 3 ** 4, "Exponentiation operator has higher precedence than multiplicative operator *");
+            assert.areEqual(2 * (3 ** 4) * 5 , 2 * 3 ** 4 * 5, "Exponentiation operator has higher precedence than multiplicative operator *");            
             assert.areEqual(2 + (3 ** 2), 2 + 3 ** 2, "Exponentiation operator has higher precedence than additive operator +");
         }
     },
+    {
+        name: "Exponentiation syntax error cases ",
+        body: function () {
+            assert.throws(function () { eval("-2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator +", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("+2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator -", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
```
Comment (ianwjhalliday) [inline test/es6/expo.js]: These first two descriptions are reversed.  Top one says unary + when the test is -, and vice versa for the second one.

File: test/es6/expo.js [hunk 5]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
             assert.areEqual(Math.pow(2,4), a, "Exponentiation assignment works as intended");
+        
             var b = -2;
             b **= -4;
             assert.areEqual(Math.pow(-2,-4), b,"Exponentiation assignment works as intended");
+
+            var c = -3;
+            var d = 0;
+            d = ++c**3;
+            assert.areEqual(Math.pow(-2,3), d,"Exponentiation assignment works as intended with preincrement operator");
```
Comment (ianwjhalliday) [inline test/es6/expo.js]: These descriptions are vague and require the reader to comprehend the test code to understand what is being tested and what is expected.  Would be better to say something clear about what is being tested, like `preincrement operator has higher precedence than exponentiation operator`.

File: test/es6/expo.js [hunk 6]
```diff
@@ -10,8 +10,8 @@ var tests = [
         name: "Exponentiation operator (**) test cases",
         body: function () {
             assert.areEqual(8 , 2 ** 3, "2 to the power of 3 is 8");
-            assert.areEqual(Math.pow(-2, -4), -2 ** -4, "Exponentiation operator ** works as intended");
-            assert.areEqual(Math.pow(-0, -4), -0 ** -4, "Exponentiation operator ** works as intended");
+            assert.areEqual(Math.pow(-2, -4), (-2) ** -4, "Exponentiation operator ** works as intended");
+            assert.areEqual(Math.pow(-0, -4), (-0) ** -4, "Exponentiation operator ** works as intended");
             assert.areEqual(Math.pow(4, 0), 4 ** -0, "Exponentiation operator ** works as intended");
```
Comment (dilijev) [inline test/es6/expo.js]: Should this test be `Math.pow(4, -0)`?

File: test/es6/expo.js [hunk 7]
```diff
@@ -10,8 +10,8 @@ var tests = [
         name: "Exponentiation operator (**) test cases",
         body: function () {
             assert.areEqual(8 , 2 ** 3, "2 to the power of 3 is 8");
-            assert.areEqual(Math.pow(-2, -4), -2 ** -4, "Exponentiation operator ** works as intended");
-            assert.areEqual(Math.pow(-0, -4), -0 ** -4, "Exponentiation operator ** works as intended");
+            assert.areEqual(Math.pow(-2, -4), (-2) ** -4, "Exponentiation operator ** works as intended");
+            assert.areEqual(Math.pow(-0, -4), (-0) ** -4, "Exponentiation operator ** works as intended");
             assert.areEqual(Math.pow(4, 0), 4 ** -0, "Exponentiation operator ** works as intended");
             assert.areEqual(Math.pow(0, -0), 0 ** -0, "Exponentiation operator ** works as intended");
             assert.areEqual(Math.pow(0, -0), 0 ** -0, "Exponentiation operator ** works as intended");
```
Comment (dilijev) [inline test/es6/expo.js]: These two tests (line 16 and 17) look identical?

File: test/es6/expo.js (hunk 1 above)
Comment (dilijev) [inline test/es6/expo.js]: Agreed with @ianwjhalliday

File: lib/Parser/perrors.h [hunk 8]
```diff
@@ -94,3 +94,4 @@ LSC_ERROR_MSG(1083, ERRBadAwait, "'await' expression not allowed in this context
 LSC_ERROR_MSG(1084, ERRGetterMustHaveNoParameters, "Getter functions must have no parameters")
 LSC_ERROR_MSG(1085, ERRFuncRefFormalNotSupportedInParamScope, "Formals cannot contain functions definitions that reference them")
 LSC_ERROR_MSG(1086, ERREvalNotSupportedInParamScope, "'eval' is not allowed in the default initializer")
+LSC_ERROR_MSG(1087, ERRInvalidUseofExponentiationOperator, "Invalid Unary operator on the left hand side of exponentiation (**) operator")
```
Comment (ianwjhalliday) [inline lib/Parser/perrors.h]: nit: unary should be all lowercase

File: lib/Parser/Parse.cpp [hunk 9]
```diff
@@ -7821,9 +7821,17 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
                     }
                 }
             }
-            else if (nop == knopEllipsis && !fAllowEllipsis)
+            else if (nop == knopEllipsis )
```
Comment (dilijev) [inline lib/Parser/Parse.cpp]: nit: there's an extra space before the close paren

File: lib/Parser/Parse.cpp (hunk 2 above)
Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: Noooooooooooooooooo

That's kind of amazing actually

Comment (dilijev) [inline lib/Parser/Parse.cpp]: Trollface indeed.

Side note, I can't actually find an environment where that code snippet above works currently. Maybe after this change we'll be the only one?

Comment (dilijev) [inline lib/Parser/Parse.cpp]: Alternate nit on this line: "left handside" should be "left hand side" to match below or "left-hand size" if we're being pedantic.

File: test/es6/expo.js [hunk 10]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
+    {
+        name: "Exponentiation syntax error cases ",
+        body: function () {
+            assert.throws(function () { eval("-2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator +", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("+2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator -", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("delete 2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator delete", "Invalid Unary operator on the left hand side of exponentiation (**) operator");
+            assert.throws(function () { eval("!2**3"); }, SyntaxError, "Exponentiation operator throws if the left hand side is an unary operator !", "Invalid Unary operator on the left hand side of exponentiation (**) operator");            
+        }
+    },
+
```
Comment (dilijev) [inline test/es6/expo.js]: nit: remove extra blank line

**Comment by dilijev** [comment]
Otherwise LGTM

File: test/es6/expo.js [hunk 11]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
+            assert.areEqual(Math.pow(-3,3), d,"Exponentiation assignment works as intended with postincrement operator");
+
+            c = -3;
+            d = 2**++c;
+            assert.areEqual(Math.pow(2,-2), d,"Exponentiation assignment works as intended with preincrement operator on the right hand side");
+
+            c = -3;
+            d = 2**c++;
+            assert.areEqual(Math.pow(2,-3), d,"Exponentiation assignment works as intended with postincrement operator on  the right hand side");
+
```
Comment (dilijev) [inline test/es6/expo.js]: nit: extra blank line

File: test/es6/expo.js [hunk 12]
```diff
@@ -34,18 +34,52 @@ var tests = [
...
             assert.areEqual(Math.pow(-2,-4), b,"Exponentiation assignment works as intended");
+
+            var c = -3;
+            var d = 0;
+            d = ++c**3;
+            assert.areEqual(Math.pow(-2,3), d,"Exponentiation assignment works as intended with preincrement operator");
+  
+            c = -3;
+            d = --c**3;
+            assert.areEqual(Math.pow(-4,3), d,"Exponentiation assignment works as intended with predecrement operator");
```
Comment (dilijev) [inline test/es6/expo.js]: nit: add a space after comma and before message in all of the above tests

File: lib/Parser/Parse.cpp (hunk 2 above)
Comment (akroshg) [inline lib/Parser/Parse.cpp]: ah nice @ljharb.

Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: @dilijev it works in ch with -es7exponentiationoperator specified.  Did you mean current stable release environments don't have that working?

Comment (dilijev) [inline lib/Parser/Parse.cpp]: Yeah that's what I meant.

**Pull request merged**